### PR TITLE
[DependencyInjection] Doc for Allow to choose an index for service locators

### PR DESCRIPTION
**Pull request by XuruDragon**
Here is the doc for the new feature implemented in : [symfony/symfony#30348](https://github.com/symfony/symfony/pull/30348) that follow the feature [symfony/symfony#30257](https://github.com/symfony/symfony/pull/30257)

### CODE REVIEW

File: service_container/service_subscribers_locators.rst [hunk 1]
```diff
@@ -372,6 +372,218 @@ will share identical locators amongst all the services referencing them::
...
 
+Tagged Services Locator Collection with Index
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+If you want to retrieve a specific service within the injected service collector
+you can use the ``index_by`` and ``default_index_method`` options of the argument
+in combination with ``!tagged_locator`` to define an index.
+
+In the following example, all services tagged with ``app.handler`` are passed as
+first constructor argument to ``App\Handler\HandlerCollection`,
```
Comment (OskarStark) [inline service_container/service_subscribers_locators.rst]:
```suggestion
first constructor argument to ``App\Handler\HandlerCollection``,
```

File: service_container/service_subscribers_locators.rst [hunk 2]
```diff
@@ -372,6 +372,218 @@ will share identical locators amongst all the services referencing them::
...
+.. code-block:: php
+
+    // src/Handler/HandlerCollection.php
+    namespace App\Handler;
+
+    use Symfony\Component\DependencyInjection\ServiceLocator;
+
+    class HandlerCollection
+    {
+        public function __construct(ServiceLocator $serviceLocator)
```
Comment (OskarStark) [inline service_container/service_subscribers_locators.rst]:
```suggestion
        public function __construct(ServiceLocator $locator)
```

File: service_container/service_subscribers_locators.rst [hunk 3]
```diff
@@ -372,6 +372,218 @@ will share identical locators amongst all the services referencing them::
...
+        {
+            $handler_two = $serviceLocator->get('handler_two'):
```
Comment (OskarStark) [inline service_container/service_subscribers_locators.rst]:
```suggestion
            $handlerTwo = $locator->get('handler_two'):
```

**Review by OskarStark — CHANGES_REQUESTED** [review]

File: service_container/service_subscribers_locators.rst [hunk 4]
```diff
@@ -372,6 +372,218 @@ will share identical locators amongst all the services referencing them::
...
+        $container->register(App\Handler\Two::class)
+            ->addTag('app.handler', ['key' => 'handler_two']);
+
+        $container->register(App\Handler\HandlerCollection::class)
+            // inject all services tagged with app.handler as first argument
+            ->addArgument(new ServiceLocatorArgument(new TaggedIteratorArgument('app.handler', 'key')));
+
+After compilation the ``HandlerCollection`` to retrieve a specific service by it's ``key`` attribute
+from the service locator injected, we just have to do ``$serviceLocator->get('handler_two');`` to
+retrieve the ``handler_two`` handler :
```
Comment (OskarStark) [inline service_container/service_subscribers_locators.rst]:
```suggestion
retrieve the ``handler_two`` handler::
```

File: service_container/service_subscribers_locators.rst [hunk 5]
```diff
@@ -372,6 +372,218 @@ will share identical locators amongst all the services referencing them::
...
+
```
Comment (OskarStark) [inline service_container/service_subscribers_locators.rst]:
```suggestion

```

File: service_container/service_subscribers_locators.rst [hunk 6]
```diff
@@ -372,6 +372,218 @@ will share identical locators amongst all the services referencing them::
...
+.. code-block:: php
```
Comment (OskarStark) [inline service_container/service_subscribers_locators.rst]:
```suggestion

```

File: service_container/service_subscribers_locators.rst [hunk 7]
```diff
@@ -372,6 +372,218 @@ will share identical locators amongst all the services referencing them::
...
+        }
+    }
+
+
+.. tip::
+
+    You can omit the ``index_attribute_name`` attribute, by implementing a static
+    method ``getDefaultIndexAttributeName`` to the handler.
+
+    Based on the previous example ``App\Handler\One`` should look like this:
```
Comment (OskarStark) [inline service_container/service_subscribers_locators.rst]:
```suggestion
Based on the previous example ``App\Handler\One`` should look like this::
```

File: service_container/service_subscribers_locators.rst [hunk 8]
```diff
@@ -372,6 +372,218 @@ will share identical locators amongst all the services referencing them::
...
+
```
Comment (OskarStark) [inline service_container/service_subscribers_locators.rst]:
```suggestion

```

File: service_container/service_subscribers_locators.rst [hunk 9]
```diff
@@ -372,6 +372,218 @@ will share identical locators amongst all the services referencing them::
...
+    .. code-block:: php
```
Comment (OskarStark) [inline service_container/service_subscribers_locators.rst]:
```suggestion

```

File: service_container/service_subscribers_locators.rst [hunk 10]
```diff
@@ -372,6 +372,218 @@ will share identical locators amongst all the services referencing them::
...
+            // config/services.php
+            $container->register(App\Handler\One::class)
+                ->addTag('app.handler', ['priority' => 20]);
+
+            // ...
+
+    You also can define the name of the static method to implement on each service
+    with the ``default_index_method`` attribute on the argument.
+
+    Based on the previous example ``App\Handler\One`` should look like :
```
Comment (OskarStark) [inline service_container/service_subscribers_locators.rst]:
```suggestion
    Based on the previous example ``App\Handler\One`` should look like::
```

File: service_container/service_subscribers_locators.rst [hunk 11]
```diff
@@ -372,6 +372,218 @@ will share identical locators amongst all the services referencing them::
...
+
```
Comment (OskarStark) [inline service_container/service_subscribers_locators.rst]:
```suggestion

```

File: service_container/service_subscribers_locators.rst [hunk 12]
```diff
@@ -372,6 +372,218 @@ will share identical locators amongst all the services referencing them::
...
+    .. code-block:: php
```
Comment (OskarStark) [inline service_container/service_subscribers_locators.rst]:
```suggestion

```

File: service_container/service_subscribers_locators.rst [hunk 13]
```diff
@@ -372,6 +372,218 @@ will share identical locators amongst all the services referencing them::
...
+    use Symfony\Component\DependencyInjection\ServiceLocator;
+
+    class HandlerCollection
+    {
+        public function __construct(ServiceLocator $locator)
+        {
+            $handlerTwo = $locator->get('handler_two'):
+        }
+    }
+
```
Comment (OskarStark) [inline service_container/service_subscribers_locators.rst]:
```suggestion

```

File: service_container/service_subscribers_locators.rst [hunk 14]
```diff
@@ -372,6 +372,218 @@ will share identical locators amongst all the services referencing them::
...
+        class One
+        {
+            public static function getDefaultIndexName(): string
+            {
+                return 'handler_one';
+            }
+        }
+
+    And the configuration:
```
Comment (OskarStark) [inline service_container/service_subscribers_locators.rst]:
```suggestion
And the configuration:
```

File: service_container/service_subscribers_locators.rst (hunk 7 above)
Comment (OskarStark) [reply service_container/service_subscribers_locators.rst]: From this point the indention of the following lines is to much!

**Review by OskarStark — CHANGES_REQUESTED** [review]

File: service_container/service_subscribers_locators.rst (hunk 7 above)
Comment (XuruDragon) [reply service_container/service_subscribers_locators.rst]: ok

File: service_container/service_subscribers_locators.rst (hunk 10 above)
Comment (XuruDragon) [reply service_container/service_subscribers_locators.rst]: why the double "::" at the end ?

File: service_container/service_subscribers_locators.rst (hunk 6 above)
Comment (XuruDragon) [reply service_container/service_subscribers_locators.rst]: why remove this one ?  my php code would not be correctly parse and show if we remove that line

Comment (OskarStark) [reply service_container/service_subscribers_locators.rst]: Thats why you should end the paragraph before with `::`

```
Here you can see my funky php highlighted::

    echo 'foo';
```

works

Comment (XuruDragon) [reply service_container/service_subscribers_locators.rst]: oh :O 
Ok I did'nt know that possibility , thanks I'll make the changes

File: service_container/service_subscribers_locators.rst (hunk 10 above)
Comment (OskarStark) [reply service_container/service_subscribers_locators.rst]: it is the default code block for php, so the following code block could be removed, see #11129 for example

**Comment by XuruDragon** [comment]
Done, thanks @OskarStark

File: service_container/service_subscribers_locators.rst [hunk 15]
```diff
@@ -372,6 +372,209 @@ will share identical locators amongst all the services referencing them::
...
+
+    use Symfony\Component\DependencyInjection\ServiceLocator;
+
+    class HandlerCollection
+    {
+        public function __construct(ServiceLocator $locator)
+        {
+            $handlerTwo = $locator->get('handler_two'):
+        }
+    }
```
Comment (OskarStark) [inline service_container/service_subscribers_locators.rst]: Please add a blank line

File: service_container/service_subscribers_locators.rst [hunk 16]
```diff
@@ -372,6 +372,209 @@ will share identical locators amongst all the services referencing them::
...
+.. tip::
+
+You can omit the ``index_attribute_name`` attribute, by implementing a static
```
Comment (OskarStark) [inline service_container/service_subscribers_locators.rst]: You need to indent this paragraph to be “in” the tip

File: service_container/service_subscribers_locators.rst [hunk 17]
```diff
@@ -372,6 +372,209 @@ will share identical locators amongst all the services referencing them::
...
+    // src/Handler/One.php
+    namespace App\Handler;
+
+    class One
+    {
+        public static function getDefaultIndexName(): string
+        {
+            return 'handler_one';
+        }
+    }
```
Comment (OskarStark) [inline service_container/service_subscribers_locators.rst]: Please add a blank line

File: service_container/service_subscribers_locators.rst [hunk 18]
```diff
@@ -372,6 +372,209 @@ will share identical locators amongst all the services referencing them::
...
+    // src/Handler/One.php
+    namespace App\Handler;
+
+    class One
+    {
+        public static function someFunctionName(): string
+        {
+            return 'handler_one';
+        }
+    }
```
Comment (OskarStark) [inline service_container/service_subscribers_locators.rst]: Please add a blank line

File: service_container/tags.rst [hunk 19]
```diff
@@ -562,3 +562,216 @@ application handlers::
...
+
+Tagged Services Collection with Index
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+If you want to retrieve a specific service within the injected collection
+you can use the ``index_by`` and ``default_index_method`` options of the argument
+in combination with ``!tagged``.
+
+In the following example, all services tagged with ``app.handler`` are passed as
+first constructor argument to ``App\Handler\HandlerCollection`,
```
Comment (OskarStark) [inline service_container/tags.rst]: One ‘ missing

File: service_container/tags.rst [hunk 20]
```diff
@@ -562,3 +562,216 @@ application handlers::
...
+            ->addTag('app.handler', ['key' => 'handler_two']);
+
+        $container->register(App\Handler\HandlerCollection::class)
+            // inject all services tagged with app.handler as first argument
+            ->addArgument(new TaggedIteratorArgument('app.handler', 'key'));
+
+After compilation the ``HandlerCollection`` is able to iterate over your
+application handlers. To retrieve a specific service by it's ``key`` attribute
+from the iterator, we can use ``iterator_to_array`` and retrieve the ``handler_two``:
+to get an array and then retrieve the ``handler_two`` handler :
```
Comment (OskarStark) [inline service_container/tags.rst]: Please use :: and remove php code block

**Review by OskarStark — CHANGES_REQUESTED** [review]

File: service_container/tags.rst (hunk 20 above)
Comment (XuruDragon) [reply service_container/tags.rst]: I've change I've rebase on the wrong branch, and I did this modification in the right PR => https://github.com/symfony/symfony-docs/pull/11009

File: service_container/service_subscribers_locators.rst [hunk 21]
```diff
@@ -372,6 +372,212 @@ will share identical locators amongst all the services referencing them::
...
+    {
+        public function __construct(ServiceLocator $locator)
+        {
+            $handlerTwo = $locator->get('handler_two'):
+        }
+    }
+
+.. tip::
+
+    You can omit the ``index_attribute_name`` attribute, by implementing a static
```
Comment (OskarStark) [inline service_container/service_subscribers_locators.rst]: This indention is ok

File: service_container/service_subscribers_locators.rst [hunk 22]
```diff
@@ -372,6 +372,212 @@ will share identical locators amongst all the services referencing them::
...
+    method ``getDefaultIndexAttributeName`` to the handler.
+
+    Based on the previous example ``App\Handler\One`` should look like this::
```
Comment (OskarStark) [inline service_container/service_subscribers_locators.rst]: From here on no indention! As it does not belong to the tip, it’s a new paragraph

File: service_container/tags.rst (hunk 19 above)
Comment (OskarStark) [reply service_container/tags.rst]: Stil missing

File: service_container/service_subscribers_locators.rst (hunk 22 above)
Comment (XuruDragon) [reply service_container/service_subscribers_locators.rst]: no It's part of the tips too, it"s the exemple of wha the tips says. 
same as the rest.

Comment (OskarStark) [reply service_container/service_subscribers_locators.rst]: Ah ok then 👍🏻

**Comment by OskarStark** [comment]
The code is merged so I removed the label.

**Review by OskarStark — APPROVED** [review]
I left a lot of comments which were all fixed.
I think this one is good to merge.

**Comment by XuruDragon** [comment]
Thanks @oskarstark

**Comment by OskarStark** [comment]
I think we are good to go here.

Note for the mergers: this must go in `4.3` branch ❗️

**Comment by javiereguiluz** [comment]
It took us some time ... but this is finally merged! @XuruDragon thanks a lot for your contribution.

**Comment by XuruDragon** [comment]
> It took us some time ... but this is finally merged! @XuruDragon thanks a lot for your contribution.

No problem, thanks to you to finally merge it ;)

**Pull request merged**